### PR TITLE
feat(calm-hub): add versioned document endpoints (Issue #2982)

### PR DESCRIPTION
**Pull request by 101Steeps**
## Description
Implements the spec described in [Issue #2982](https://github.com/finos/architecture-as-code/issues/2982) as a part of [Issue #2981 ](https://github.com/finos/architecture-as-code/issues/2981)

**Excerpt from  [Issue #2982](https://github.com/finos/architecture-as-code/issues/2982l):**
Add a first-class Document resource to CALM Hub so a document (Markdown + structured frontmatter) can be created, retrieved, and updated by document type. This is the storage/API foundation for the Documents PoC and implements the decision in https://github.com/finos/architecture-as-code/issues/2791 (documents are first-class, not decorators) using the document shape agreed in https://github.com/finos/architecture-as-code/issues/2866.

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation update
- [ ] 🎨 Code style/formatting changes
- [ ] ♻️ Refactoring (no functional changes)
- [ ] ⚡ Performance improvements
- [x] ✅ Test additions or updates
- [ ] 🔧 Chore (maintenance, dependencies, CI, etc.)

## Affected Components

- [ ] CLI (`cli/`)
- [ ] Schema (`calm/`)
- [ ] CALM AI (`calm-ai/`)
- [x] CALM Hub (`calm-hub/`)
- [ ] CALM Hub UI (`calm-hub-ui/`)
- [ ] CALM Server (`calm-server/`)
- [ ] CALM Widgets (`calm-widgets/`)
- [ ] Documentation (`docs/`)
- [ ] Shared (`shared/`)
- [ ] VS Code Extension (`calm-plugins/vscode/`)
- [ ] Dependencies
- [ ] CI/CD

## Testing

- [x] I have tested my changes locally
- [x] I have added/updated unit tests
- [x] All existing tests pass

- `npm run build` passed.
- `npm test -- --coverage` passed.
- `npm run lint` passed with 0 errors. The command reported 11 warnings.
- `./mvnw clean install -Ddependency-check.skip=true` passed.
- `calm-hub` unit tests passed: 2,772 tests.
- `calm-hub` integration tests passed: 537 tests.
- JaCoCo coverage checks passed for `calm-hub`.
- Swagger testing passed. The document create endpoint returned `201 Created`.
- The Swagger response included the versioned `Location` header.
- Swagger testing confirmed a successful document create request.
- The endpoint returned `201 Created`.
- The response included a versioned `Location` header.
- The audit log recorded `CREATE`, `DOCUMENT`, `swaggeraudit`, document ID `1`, version `1.0.0`, and `SUCCESS`.

**Swagger / Quarkus Testing**
_Swagger endpoint discovery: Document endpoints for listing, creating, retrieving, and versioning narrative documents._
image

_Swagger manual test: Creating a `knowledge` document returned `201 Created` with a versioned `Location` header._
image

## Checklist
- [x] My commits follow the [conventional commit format](https://www.conventionalcommits.org/)
- [x] I have updated documentation if necessary
- [x] I have added tests for my changes (if applicable)
- [x] My changes follow the project's coding standards

### CODE REVIEW

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+@Tag(name = "Storage API", description = "Numeric-ID based CALM storage endpoints")
+@Path("/api/calm/namespaces")
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+public class DocumentResource {
+    /** Temporary Java-side POC contract pending a published shared contract. */
+    private final DocumentService documentService;
+    public DocumentResource(DocumentService documentService) { this.documentService = documentService; }
+    @GET @Path("{namespace}/documents/{documentType}") @PermissionsAllowed(CalmHubScopes.READ)
```
Comment (LeighFinegold) [inline calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java]: The formatting here is very dense compared to the rest of the codebase. Here's the style to follow -[`PatternResource.java` L55–76](https://github.com/finos/architecture-as-code/blob/8277e874d774/calm-hub/src/main/java/org/finos/calm/resources/PatternResource.java#L55-L76) shows a single endpoint with:

- Each annotation (`@GET`, `@Path`, `@Produces`, `@Operation`, `@PermissionsAllowed`) on its own line
- `@Operation` with both `summary` and `description` fields
- Method params on separate lines
- Try/catch bodies expanded, not single-line
- Logger field ([L44](https://github.com/finos/architecture-as-code/blob/8277e874d774/calm-hub/src/main/java/org/finos/calm/resources/PatternResource.java#L44)) and logging in each catch block ([L73](https://github.com/finos/architecture-as-code/blob/8277e874d774/calm-hub/src/main/java/org/finos/calm/resources/PatternResource.java#L73))

Could you reformat the new resource to match?

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        catch (DocumentVersionExistsException e) { return Response.status(409).entity("Document version already exists").build(); }
+    }
+    private Response validate(String type, CreateDocumentRequest request) {
+        if (!DOCUMENT_TYPES.contains(type)) return Response.status(400).entity("Unsupported document type").build();
+        if (request == null || !hasMappingFrontmatter(request.getDocumentMarkdown())) return Response.status(400).entity("documentMarkdown must begin with YAML mapping frontmatter").build();
+        return null;
+    }
+    private boolean hasMappingFrontmatter(String markdown) {
+        if (markdown == null) return false;
+        java.util.regex.Matcher frontmatter = java.util.regex.Pattern
```
Comment (LeighFinegold) [inline calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java]: The frontmatter regex is compiled on every call - `Pattern.compile()` is relatively expensive and the compiled `Pattern` is thread-safe and immutable, so it only needs to happen once. 

I noticed [`VERSION_PATTERN` in `MongoDocumentStore`](https://github.com/finos/architecture-as-code/pull/3002/files#diff-MongoDocumentStore) and [`NitriteDocumentStore`](https://github.com/finos/architecture-as-code/pull/3002/files#diff-NitriteDocumentStore) are already done correctly as static fields - same approach should be applied here.

File: calm-hub/src/main/java/org/finos/calm/store/producer/DocumentStoreProducer.java [hunk 3]
```diff
@@ -0,0 +1,3 @@
+package org.finos.calm.store.producer;
```
Comment (LeighFinegold) [inline calm-hub/src/main/java/org/finos/calm/store/producer/DocumentStoreProducer.java:1]: @101Steeps  - per ask, did a first scan through before making PR beyond draft. 

Think you need to do run the file formatter :)

Have you had a chance to read through [`calm-hub/AGENTS.md`](https://github.com/finos/architecture-as-code/blob/8277e874d774/calm-hub/AGENTS.md)? It covers the codebase conventions we follow. Might be worth checking your agent has this.

File: calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java [hunk 4]
```diff
@@ -13,6 +15,8 @@ public class ResourceValidationConstants {
     public static final String DOMAIN_MESSAGE = "domain name must match pattern '^[A-Za-z0-9-]+$'";
     public static final String VERSION_REGEX = "^(0|[1-9][0-9]*)[-.]?(0|[1-9][0-9]*)[-.]?(0|[1-9][0-9]*)$";
     public static final String VERSION_MESSAGE = "version must match pattern '^(0|[1-9][0-9]*)[-.]?(0|[1-9][0-9]*)[-.]?(0|[1-9][0-9]*)$'";
+    /** Temporary POC types pending a shared Java-consumable document contract. */
+    public static final Set<String> DOCUMENT_TYPES = Set.of("pattern", "architecture", "interface", "flow", "control", "schema", "timeline", "adr");
```
Comment (LeighFinegold) [inline calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java]: This is the allow-list for `/documents/{documentType}` - I don't think this is the right list. These are the existing CALM resource types which already have their own dedicated endpoints (`/patterns`, `/architectures`, etc.). This should be narrative document types like `knowledge`, `sad` — no?

Probably worth calling it NARRATIVE_DOCUMENT_TYPES although variable naming was never my forte

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java (hunk 1 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java]: Updated 'DocumentResource' to follow style in PatternResource

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java (hunk 2 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java]: Moved the frontmatter regex to the static fields so 'DocumentResource' only compiles the pattern once and resuses for each request

File: calm-hub/src/main/java/org/finos/calm/store/producer/DocumentStoreProducer.java (hunk 3 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/store/producer/DocumentStoreProducer.java:1]: Changed the files to match the repo and calm hub formatting conventions. I've read through the Agents.md for Calm-hub now and made sure its now included in the review step.

File: calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java (hunk 4 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java]: Added 'NARRATIVE_DOCUMENT_TYPES' (happy to use this variable name), with supported values 'knowledge' and 'sad'. Document endpoints reject unsupported narrative document types. I've added additional test evidence to the PR description to show Swagger and the Quarkus audit log for this

File: calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    @ConfigProperty(name = "calm.database.mode", defaultValue = "mongo")
+    String databaseMode;
+
+    public MongoDocumentIndexStep(MongoDatabase database) {
+        this.database = database;
+    }
+
+    @Override
+    public int fromVersion() {
+        return 14;
```
Comment (rocketstack-matt) [inline calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java:35]: `fromVersion() == 14` collides with `MongoLayoutFormatMigrationStep`, which merged to `main` at `fromVersion() == 14` after this branch diverged. `SchemaMigrationRunner.indexByFromVersion()` throws `IllegalStateException: Duplicate SchemaMigrationStep for fromVersion 14` when both beans are on the classpath together — calm-hub won't start once this merges. This PR's own CI passed only because it ran before the colliding step existed on main.

Needs the next actually-free `fromVersion` after rebasing, plus a matching `LATEST_SCHEMA_VERSION` bump in `calm-hub/mongo/init-mongo.js` (currently set to 15 here, which is also now stale).

File: calm-hub/src/main/java/org/finos/calm/store/mongo/MongoDocumentStore.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+import java.util.Comparator;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@LookupIfProperty(name = "calm.database.mode", stringValue = "mongo", lookupIfMissing = true)
+@ApplicationScoped
+@Typed(MongoDocumentStore.class)
+public class MongoDocumentStore implements DocumentStore {
+    private static final String COLLECTION = "documents";
```
Comment (rocketstack-matt) [inline calm-hub/src/main/java/org/finos/calm/store/mongo/MongoDocumentStore.java:36]: This store uses one document per `(namespace, documentType)` holding an array of documents, each with an embedded dash-encoded (`"1-0-0"`) `versions` map. `calm-hub/decisions/0001-versioned-artefact-storage.md` moved every other versioned type (Architecture, Pattern, Flow, Standard, Interface, Timeline, ADR) off exactly this shape, specifically to avoid Mongo's 16MB/document limit and Nitrite's full-document read-modify-write lock contention. ADR 0002 also dropped dash-encoded version keys in favor of dot-separated. Narrative documents (full Markdown SADs) are arguably *more* likely to hit the 16MB ceiling than JSON architectures, and no regression test analogous to `MongoDocumentSizeLimitIntegration` was added here.

This should follow the header + `documentVersions` collection pattern via `MongoVersionSplitMigration` (ADR 0003) instead of introducing a ninth instance of the shape the rest of the codebase already migrated away from. Same issue applies to `NitriteDocumentStore`.

File: calm-hub/src/main/java/org/finos/calm/store/nitrite/NitriteDocumentStore.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+import java.util.*;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@LookupIfProperty(name = "calm.database.mode", stringValue = "standalone")
+@ApplicationScoped
+@Typed(NitriteDocumentStore.class)
+public class NitriteDocumentStore implements DocumentStore {
```
Comment (rocketstack-matt) [inline calm-hub/src/main/java/org/finos/calm/store/nitrite/NitriteDocumentStore.java:29]: Same shape/versioning concern as `MongoDocumentStore` (see that file's comment) — this should use `NitriteVersionDocumentStore` rather than a hand-rolled embedded array with a dash-encoded versions map.

Separately: parameter/variable names throughout this class (`n`, `t`, `r`, `d`, `v`, `vs`, `k`) are much denser than the codebase convention — compare `NitriteStandardStore`. This is the same "too dense" issue already flagged and fixed in `DocumentResource.java`; worth applying the same pass here.

**Review by rocketstack-matt — DISMISSED** [review]
Two blocking issues: a duplicate schema-migration version that will crash calm-hub startup once merged, and a storage shape that reintroduces the exact problem ADR 0001/0002 already fixed for every other versioned type.

File: calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java (hunk 5 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java]: Fixed in 66d673e7. The layout migration retains transition 14 → 15, the document migration uses 15 → 16. LATEST_SCHEMA_VERSION in init-mongo.js is also 16.

File: calm-hub/src/main/java/org/finos/calm/store/mongo/MongoDocumentStore.java (hunk 6 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/store/mongo/MongoDocumentStore.java]: Fixed in 66d673e7. Narrative documents now use a header per (namespace, documentType, documentId) and separate documentVersions records, following ADR 0001.

Existing embedded data is migrated through MongoVersionSplitMigration, reusing the shared helper described in ADR 0003. Migrated versions and new writes use canonical dot-separated fields as required by ADR 0002. The unique indexes include documentType.

Added a narrative case to MongoDocumentSizeLimitIntegration that writes approximately 24 MB across 12 versions and reads each back with its Markdown unchanged. This passed locally and verifies that aggregate history can exceed the old ceiling .Nitrite now uses the same header/version storage shape.

File: calm-hub/src/main/java/org/finos/calm/store/nitrite/NitriteDocumentStore.java (hunk 7 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/store/nitrite/NitriteDocumentStore.java:26]: Fixed in 66d673e7. NitriteDocumentStore now delegates to NitriteVersionDocumentStore, with separate headers and canonical version records. Helper instances are cached per document type so requests share the lock used for duplicate detection and insertion.

Replaced the abbreviated parameter and variable names with descriptive names. Added tests for migration failures and retries, conflicting aliases, and cleanup when the first version cannot be written. String and document content now share the same write and cleanup implementation.

**Review by markscott-ms — CHANGES_REQUESTED** [review]
There's a few aspects of this I'm not happy with.

1. The change to use BigInteger and an additional version parsing class. The reason for supporting versioning elements larger than 2^31 is not explained, and the separate class potentially breaks handling because parts of the code still use the `Semver` record. If values above 2^31 really are required (why?), the `Semver` record could have been changed.
2. Reformatting of several areas hinders (manual) review
3. A significant change to MongoVersionSplitMigration

I'm going to go through the code in detail but wanted to raise these concerns whilst I do that.

File: calm-hub/decisions/0003-shared-version-store-helper.md [hunk 8]
```diff
@@ -133,8 +133,9 @@ rules can't drift apart between them.
   `1.10.0` before `1.9.0`, so ordering had to become explicit once versions
   were rows rather than map keys. Deliberately *not* a change to
   `VersionKeySelector`, which keeps parsing dashes for Control — see
-  [ADR 0002](0002-version-key-encoding.md). It delegates to the existing
-  `Semver` record rather than parsing versions a second time.
+  [ADR 0002](0002-version-key-encoding.md). It originally delegated to the
+  `Semver` record. It now uses `BigInteger` components because
+  the API accepts version components above the Java integer range.
```
Comment (markscott-ms) [inline calm-hub/decisions/0003-shared-version-store-helper.md:138]: ADRs should be superseded rather than modified, particularly where the decision changes.

This isn't a decision change, but it does not fully explain why `BigInteger` has been used rather than improving the range supported by the `Semver` class, which doesn't obviously tie to the purpose of this PR of supporting documents.

File: calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+@LookupIfProperty(name = "calm.database.mode", stringValue = "mongo", lookupIfMissing = true)
+@ApplicationScoped
+public class MongoDocumentIndexStep implements SchemaMigrationStep {
+    private final MongoVersionSplitMigration migration;
+
+    @ConfigProperty(name = "calm.database.mode", defaultValue = "mongo")
+    String databaseMode;
+
+    public MongoDocumentIndexStep(MongoDatabase database) {
+        migration = new MongoVersionSplitMigration(database, "documents", "documentVersions", "documentId",
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java:19]: Documents have never been stored in Mongo or Nitrite yet... so there is nothing to migrate, only a new collection to create and an index to add to it.

File: calm-hub/src/main/java/org/finos/calm/migration/steps/MongoVersionSplitMigration.java [hunk 10]
```diff
@@ -103,6 +106,15 @@ public MongoVersionSplitMigration(MongoDatabase database, String headerCollectio
     public MongoVersionSplitMigration(MongoDatabase database, String headerCollection, String versionCollection,
                                       String idField, String arrayField, String versionsField, String resourceLabel,
                                       VersionScheme versionScheme) {
+        this(database, headerCollection, versionCollection, idField, arrayField, versionsField,
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/migration/steps/MongoVersionSplitMigration.java:109]: Documents have never been stored in Mongo or Nitrite yet... so there is nothing to migrate, only a new collection to create and an index to add to it. This means all these changes to MongoVersionSplitMigration should be unnecessary.

Am I missing something?

File: calm-hub/src/main/java/org/finos/calm/migration/steps/NitriteDocumentVersionSplitStep.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Override
+    public int fromVersion() {
+        return 15;
+    }
+
+    @Override
+    public void apply() {
+        NitriteCollection headers = database.getCollection("documents");
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/migration/steps/NitriteDocumentVersionSplitStep.java:41]: Documents have never been stored in Mongo or Nitrite yet... so there is nothing to migrate, only a new collection to create and an index to add to it.

File: calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java [hunk 12]
```diff
@@ -3,23 +3,38 @@
 import org.owasp.html.HtmlPolicyBuilder;
 import org.owasp.html.PolicyFactory;
 
+import java.util.Set;
+
 public class ResourceValidationConstants {
     public static final String NAMESPACE_REGEX = "^[A-Za-z0-9-]+(\\.[A-Za-z0-9-]+)*$";
-    public static final String NAMESPACE_MESSAGE = "namespace must match pattern '^[A-Za-z0-9-]+([.][A-Za-z0-9-]+)*$'";
-    // dashes, dots, usernames, upper/lowercase letters, numbers, and * for the public-access wildcard.
+    public static final String NAMESPACE_MESSAGE =
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java:10]: please avoid reformatting lines you haven't changed. It makes review very hard, unless reformats are in a separate commit, and called out as such in a PR description to aid reviewers to ignore them

File: calm-hub/src/main/java/org/finos/calm/security/AuditRequestFilter.java [hunk 13]
```diff
@@ -42,70 +47,66 @@
 import java.util.Map;
 import java.util.Set;
 
-import static org.finos.calm.resources.ResourceValidationConstants.STRICT_SANITIZATION_POLICY;
-
 /**
- * Captures an {@link AuditLogEntry} for every mutating (POST/PUT/DELETE) request that
- * completes with a {@code SUCCESS} (2xx) or {@code DENIED} (401/403) outcome.
+ * Captures an {@link AuditLogEntry} for every mutating (POST/PUT/DELETE) request that completes
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/security/AuditRequestFilter.java:51]: Again, reformatting makes it harder to see what has truly changed in this file

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+            description =
+                    "Returns all documents stored for the supplied namespace and narrative document"
+                            + " type.")
+    @APIResponse(responseCode = "200", description = "Documents returned")
+    @PermissionsAllowed(CalmHubScopes.READ)
+    public Response list(
+            @PathParam("namespace") @Pattern(regexp = NAMESPACE_REGEX, message = NAMESPACE_MESSAGE)
+                    String namespace,
+            @PathParam("documentType") String type) {
+        if (!NARRATIVE_DOCUMENT_TYPES.contains(type)) {
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java:73]: Duplicates logic in `validate()` that could be re-used.

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+            summary = "List narrative document versions",
+            description = "Returns the immutable versions for a narrative document.")
+    @APIResponse(responseCode = "200", description = "Versions returned")
+    @PermissionsAllowed(CalmHubScopes.READ)
+    public Response versions(
+            @PathParam("namespace") @Pattern(regexp = NAMESPACE_REGEX, message = NAMESPACE_MESSAGE)
+                    String namespace,
+            @PathParam("documentType") String type,
+            @PathParam("id") Integer id) {
+        if (!NARRATIVE_DOCUMENT_TYPES.contains(type)) {
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java:144]: Duplicates logic in `validate()` that could be re-used.

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+    @APIResponse(responseCode = "200", description = "Document Markdown returned")
+    @PermissionsAllowed(CalmHubScopes.READ)
+    public Response get(
+            @PathParam("namespace") @Pattern(regexp = NAMESPACE_REGEX, message = NAMESPACE_MESSAGE)
+                    String namespace,
+            @PathParam("documentType") String type,
+            @PathParam("id") Integer id,
+            @PathParam("version") @Pattern(regexp = VERSION_REGEX, message = VERSION_MESSAGE)
+                    String version) {
+        if (!NARRATIVE_DOCUMENT_TYPES.contains(type)) {
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java:182]: Duplicates logic in `validate()` that could be re-used.

File: calm-hub/src/main/java/org/finos/calm/store/util/SemanticVersionOrder.java [hunk 17]
```diff
@@ -13,15 +12,8 @@
...
- * <h2>Separators</h2>
- * Parsing is delegated to {@link Semver#tryParse}, so both the canonical
- * dot-separated form and the dash-encoded form order identically. ADR 0002 has
- * the new {@code <type>Versions} collections storing dots only, so dashes should
- * never reach here — but {@code VERSION_REGEX} accepts either from the API, and a
- * comparator silently demoting {@code "1-10-0"} to {@code 0.0.0} would be a
- * confusing way to find that out. Accepting both costs nothing and keeps this
- * consistent with every other version comparison in the codebase, all of which
- * already go through {@code Semver}.
+ * Accepts dot and dash separators. Components use BigInteger so versions
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/store/util/SemanticVersionOrder.java:15]: This class should be using Semver. If support for `> 2^31` is necessary, then Semver must be modified, not this class. Doing it only in this class potentially breaks all other code that is using Semver for ordering.

I'm not clear why support for `> 2^31` is necessary.

**Review by markscott-ms — CHANGES_REQUESTED** [review]

File: calm-hub/decisions/0003-shared-version-store-helper.md (hunk 8 above)
Comment (101Steeps) [reply calm-hub/decisions/0003-shared-version-store-helper.md]: I have removed the PR-specific ADR change and the separate BigInteger handling. SemanticVersionOrder now follows the shared Semver behaviour, and ADR 0003 is unchanged from main.

File: calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java (hunk 9 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/migration/steps/MongoDocumentIndexStep.java]: Agreed. The agent-assisted implementation had carried forward migration logic from an earlier iteration of this branch, effectively treating development history within the PR as though it were an existing deployed schema. I've corrected this going forward. 

I've simplified MongoDocumentIndexStep so it now only creates the two unique indexes required by the new storage model: the document header index and the documentVersions index. There is no document data transformation in this step anymore.

File: calm-hub/src/main/java/org/finos/calm/migration/steps/MongoVersionSplitMigration.java (hunk 10 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/migration/steps/MongoVersionSplitMigration.java]: Same issue as above with development history being treated as an existing schema. 
I've removed all document-specific changes from MongoVersionSplitMigration; the class is now unchanged from main.

New Documents are written directly using the header + separate version-record storage model, so the migration remains concerned only with resource types that genuinely have an older persisted layout.

File: calm-hub/src/main/java/org/finos/calm/migration/steps/NitriteDocumentVersionSplitStep.java (hunk 11 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/migration/steps/NitriteDocumentVersionSplitStep.java]: Agreed, for the same reason as Mongo above. 

I've removed NitriteDocumentVersionSplitStep and its associated migration tests entirely. Nitrite Documents now start directly with the new header + separate version-record storage model, without implying that a legacy Documents representation exists.

File: calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java (hunk 12 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/resources/ResourceValidationConstants.java]: Fixed this now, was the result of agent generated changes formatting surrounding code while making the change. 
I've restored the existing formatting to main. The final diff in this file is now limited to the Set import and the new NARRATIVE_DOCUMENT_TYPES allow-list for knowledge and sad.

File: calm-hub/src/main/java/org/finos/calm/security/AuditRequestFilter.java (hunk 13 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/security/AuditRequestFilter.java]: Same issue as above. I've restored the existing formatting and final diff now contains only the Documents-specific audit additions: the DocumentResource import, the mapping to AuditEntityType.DOCUMENT, and the "id" path-parameter mapping used by the Document endpoints.

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java (hunk 14 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java]: I've extracted that logic into validateDocumentType(...). Endpoints that only require the type check call that helper directly, while create/update continue through validate(...), which now delegates to the same helper before performing request-body validation.

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java (hunk 15 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java]: Covererd by same refactor as above, duplicated document-type check has been moved into validateDocumentType(...), and this endpoint now reuses that helper rather than maintaining its own copy of the logic.

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java (hunk 16 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java]: Covererd by same refactor as above, duplicated document-type check has been moved into validateDocumentType(...), and this endpoint now reuses that helper rather than maintaining its own copy of the logic.

File: calm-hub/src/main/java/org/finos/calm/store/util/SemanticVersionOrder.java (hunk 17 above)
Comment (101Steeps) [reply calm-hub/src/main/java/org/finos/calm/store/util/SemanticVersionOrder.java]: Agent introduced the separate BigInteger parsing while trying to cover an extreme edge case, but there's no Documents-specific requirement for version components greater than 2^31. I should've caught this before it made it to PR. 
I've removed the oversized-version behaviour and associated tests rather than changing the global version model as part of this work. SemanticVersionOrder.java itself now has no diff from main.

File: calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+            return false;
+        }
+        Matcher frontmatter = FRONTMATTER_PATTERN.matcher(markdown);
+        if (!frontmatter.find()) {
+            return false;
+        }
+        try {
+            Object parsed = new Yaml().load(frontmatter.group(1));
+            return parsed instanceof Map<?, ?> mapping && !mapping.isEmpty();
+        } catch (RuntimeException e) {
```
Comment (markscott-ms) [inline calm-hub/src/main/java/org/finos/calm/resources/DocumentResource.java:292]: `RuntimeException` is sufficient. Modern SnakeYamls have a default nesting limit of 50 meaning that `StackOverflowError` is highly unlikely (`Error`s are not subclasses of `RuntimeException`).

**Review by markscott-ms — APPROVED** [review]
All review comments addressed, no further findings
